File: artifacts/haskell/src/MySpace/MusicCatalogue/Statements/SelectAlbumFields.hs
module MySpace.MusicCatalogue.Statements.SelectAlbumFields where

import MySpace.MusicCatalogue.Prelude
import qualified Hasql.Statement as Statement
import qualified Hasql.Decoders as Decoders
import qualified Hasql.Encoders as Encoders
import qualified Data.Aeson as Aeson
import qualified Data.Vector as Vector
import qualified Hasql.Mapping.IsStatement as IsStatement
import qualified Hasql.Mapping.IsScalar as IsScalar
import qualified MySpace.MusicCatalogue.Types as Types
import qualified PostgresqlTypes as Pt

-- |
-- Parameters for the @select_album_fields@ query.
--
-- ==== SQL Template
--
-- > -- Demonstrates static query equivalent of dynamic field selection.
-- > -- Boolean flags control which fields are included in the result,
-- > -- returning NULL for fields the caller opts out of.
-- > SELECT
-- >   album.id,
-- >   CASE WHEN $include_name      THEN album.name      END AS name,
-- >   CASE WHEN $include_released  THEN album.released  END AS released,
-- >   CASE WHEN $include_format    THEN album.format    END AS format,
-- >   CASE WHEN $include_recording THEN album.recording END AS recording,
-- >   CASE WHEN $include_tracks    THEN album.tracks    END AS tracks,
-- >   CASE WHEN $include_disc      THEN album.disc      END AS disc
-- > FROM album
-- > WHERE album.id = $id
--
-- ==== Source Path
--
-- > ./queries/select_album_fields.sql
--
data SelectAlbumFields = SelectAlbumFields
  { -- | Maps to @include_name@.
    includeName :: Bool,
    -- | Maps to @include_released@.
    includeReleased :: Bool,
    -- | Maps to @include_format@.
    includeFormat :: Bool,
    -- | Maps to @include_recording@.
    includeRecording :: Bool,
    -- | Maps to @include_tracks@.
    includeTracks :: Bool,
    -- | Maps to @include_disc@.
    includeDisc :: Bool,
    -- | Maps to @id@.
    id :: Int64
  }
  deriving stock (Eq, Show)

-- | Result of the statement parameterised by 'SelectAlbumFields'.
type SelectAlbumFieldsResult = Vector.Vector SelectAlbumFieldsResultRow

-- | Row of 'SelectAlbumFieldsResult'.
data SelectAlbumFieldsResultRow = SelectAlbumFieldsResultRow
  { -- | Maps to @id@.
    id :: Int64,
    -- | Maps to @name@.
    name :: Maybe (Text),
    -- | Maps to @released@.
    released :: Maybe (Pt.Date),
    -- | Maps to @format@.
    format :: Maybe (Types.AlbumFormat),
    -- | Maps to @recording@.
    recording :: Maybe (Types.RecordingInfo),
    -- | Maps to @tracks@.
    tracks :: Maybe (Vector (Types.TrackInfo)),
    -- | Maps to @disc@.
    disc :: Maybe (Types.DiscInfo)
  }
  deriving stock (Show, Eq)

instance IsStatement.IsStatement SelectAlbumFields where
  type Result SelectAlbumFields = SelectAlbumFieldsResult

  statement = Statement.preparable sql encoder decoder
    where
      sql =
        "-- Demonstrates static query equivalent of dynamic field selection.\n\
        \-- Boolean flags control which fields are included in the result,\n\
        \-- returning NULL for fields the caller opts out of.\n\
        \SELECT\n\
        \  album.id,\n\
        \  CASE WHEN $1      THEN album.name      END AS name,\n\
        \  CASE WHEN $2  THEN album.released  END AS released,\n\
        \  CASE WHEN $3    THEN album.format    END AS format,\n\
        \  CASE WHEN $4 THEN album.recording END AS recording,\n\
        \  CASE WHEN $5    THEN album.tracks    END AS tracks,\n\
        \  CASE WHEN $6      THEN album.disc      END AS disc\n\
        \FROM album\n\
        \WHERE album.id = $7"

      encoder =
        mconcat
          [ (.includeName) >$< Encoders.param (Encoders.nonNullable (IsScalar.encoder)),
            (.includeReleased) >$< Encoders.param (Encoders.nonNullable (IsScalar.encoder)),
            (.includeFormat) >$< Encoders.param (Encoders.nonNullable (IsScalar.encoder)),
            (.includeRecording) >$< Encoders.param (Encoders.nonNullable (IsScalar.encoder)),
            (.includeTracks) >$< Encoders.param (Encoders.nonNullable (IsScalar.encoder)),
            (.includeDisc) >$< Encoders.param (Encoders.nonNullable (IsScalar.encoder)),
            (.id) >$< Encoders.param (Encoders.nonNullable (IsScalar.encoder))
          ]

      decoder =
        Decoders.rowVector do
          id <- Decoders.column (Decoders.nonNullable (IsScalar.decoder))
          name <- Decoders.column (Decoders.nullable (IsScalar.decoder))
          released <- Decoders.column (Decoders.nullable (IsScalar.decoder))
          format <- Decoders.column (Decoders.nullable (IsScalar.decoder))
          recording <- Decoders.column (Decoders.nullable (IsScalar.decoder))
          tracks <- Decoders.column (Decoders.nullable (Decoders.array (Decoders.dimension Vector.replicateM (Decoders.element (Decoders.nonNullable IsScalar.decoder)))))
          disc <- Decoders.column (Decoders.nullable (IsScalar.decoder))
          pure SelectAlbumFieldsResultRow {..}

File: artifacts/java_null/src/main/java/io/pgenie/artifacts/myspace/musiccatalogue/statements/SelectAlbumFields.java
package io.pgenie.artifacts.myspace.musiccatalogue.statements;

import java.sql.Date;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.time.*;
import java.util.ArrayList;
import java.util.List;
import io.codemine.java.postgresql.jdbc.Codec;
import io.codemine.java.postgresql.jdbc.Statement;
import io.pgenie.artifacts.myspace.musiccatalogue.types.*;

/**
 * Type-safe binding for the {@code select_album_fields} query.
 *
 * <h2>SQL Template</h2>
 *
 * <pre>{@code
 * -- Demonstrates static query equivalent of dynamic field selection.
 * -- Boolean flags control which fields are included in the result,
 * -- returning NULL for fields the caller opts out of.
 * SELECT
 *   album.id,
 *   CASE WHEN $include_name      THEN album.name      END AS name,
 *   CASE WHEN $include_released  THEN album.released  END AS released,
 *   CASE WHEN $include_format    THEN album.format    END AS format,
 *   CASE WHEN $include_recording THEN album.recording END AS recording,
 *   CASE WHEN $include_tracks    THEN album.tracks    END AS tracks,
 *   CASE WHEN $include_disc      THEN album.disc      END AS disc
 * FROM album
 * WHERE album.id = $id
 * }</pre>
 *
 * <h2>Source Path</h2> {@code ./queries/select_album_fields.sql}
 *
 * <p>
 * Generated from SQL queries using the
 * <a href="https://pgenie.io">pGenie</a> code generator.
 */
public record SelectAlbumFields(
        /**
         * Maps to {@code $include_name} in the template.
         */
        boolean includeName,
        /**
         * Maps to {@code $include_released} in the template.
         */
        boolean includeReleased,
        /**
         * Maps to {@code $include_format} in the template.
         */
        boolean includeFormat,
        /**
         * Maps to {@code $include_recording} in the template.
         */
        boolean includeRecording,
        /**
         * Maps to {@code $include_tracks} in the template.
         */
        boolean includeTracks,
        /**
         * Maps to {@code $include_disc} in the template.
         */
        boolean includeDisc,
        /**
         * Maps to {@code $id} in the template.
         */
        long id)
        implements Statement<SelectAlbumFields.Result> {
    
    // -------------------------------------------------------------------------
    // Result type
    // -------------------------------------------------------------------------
    /**
     * Result of the statement parameterised by {@link SelectAlbumFields}.
     */
    public static final class Result extends ArrayList<ResultRow> {
        Result() {}
    }

    /**
     * Row of {@link Result}.
     */
    public record ResultRow(
            /**
             * Maps to the {@code id} result-set column.
             */
            long id,
            /**
             * Maps to the {@code name} result-set column. Nullable.
             */
            String name,
            /**
             * Maps to the {@code released} result-set column. Nullable.
             */
            LocalDate released,
            /**
             * Maps to the {@code format} result-set column. Nullable.
             */
            AlbumFormat format,
            /**
             * Maps to the {@code recording} result-set column. Nullable.
             */
            RecordingInfo recording,
            /**
             * Maps to the {@code tracks} result-set column. Nullable.
             */
            List<TrackInfo> tracks,
            /**
             * Maps to the {@code disc} result-set column. Nullable.
             */
            DiscInfo disc) {}
    
    // -------------------------------------------------------------------------
    // Statement implementation
    // -------------------------------------------------------------------------
    @Override
    public String sql() {
        return """
               -- Demonstrates static query equivalent of dynamic field selection.
               -- Boolean flags control which fields are included in the result,
               -- returning NULL for fields the caller opts out of.
               SELECT
                 album.id,
                 CASE WHEN ?      THEN album.name      END AS name,
                 CASE WHEN ?  THEN album.released  END AS released,
                 CASE WHEN ?    THEN album.format    END AS format,
                 CASE WHEN ? THEN album.recording END AS recording,
                 CASE WHEN ?    THEN album.tracks    END AS tracks,
                 CASE WHEN ?      THEN album.disc      END AS disc
               FROM album
               WHERE album.id = ?
               """;
    }

    @Override
    public void bindParams(PreparedStatement ps) throws SQLException {
        Codec.BOOL.bind(ps, 1, this.includeName());
        Codec.BOOL.bind(ps, 2, this.includeReleased());
        Codec.BOOL.bind(ps, 3, this.includeFormat());
        Codec.BOOL.bind(ps, 4, this.includeRecording());
        Codec.BOOL.bind(ps, 5, this.includeTracks());
        Codec.BOOL.bind(ps, 6, this.includeDisc());
        Codec.INT8.bind(ps, 7, this.id());
    }

    @Override
    public boolean returnsRows() {
        return true;
    }

    @Override
    public Result decodeResultSet(ResultSet rs) throws SQLException {
        Result output = new Result();
        int row = 0;
        
        while (rs.next()) {
            long idCol = Codec.INT8.decodeNonNullable(rs, row, 1);
            String nameCol = Codec.TEXT.decodeNullable(rs, row, 2);
            LocalDate releasedCol = Codec.DATE.decodeNullable(rs, row, 3);
            AlbumFormat formatCol = AlbumFormat.CODEC.decodeNullable(rs, row, 4);
            RecordingInfo recordingCol = RecordingInfo.CODEC.decodeNullable(rs, row, 5);
            List<TrackInfo> tracksCol = TrackInfo.CODEC.inDim().decodeNullable(rs, row, 6);
            DiscInfo discCol = DiscInfo.CODEC.decodeNullable(rs, row, 7);

            output.add(new ResultRow(idCol, nameCol, releasedCol, formatCol, recordingCol, tracksCol, discCol));
            row++;
        }

        return output;
    }

    @Override
    public SelectAlbumFields.Result decodeAffectedRows(long affectedRows) {
        throw new UnsupportedOperationException();
    }
}
